pkg/protocol/engine/congestioncontrol/scheduler/drr/drrbuffer.go

```diff
@@ -0,0 +1,321 @@
+
+// BufferQueue represents a buffer of IssuerQueue.
+type BufferQueue struct {
+	maxBuffer int
+
+	activeIssuers *shrinkingmap.ShrinkingMap[iotago.AccountID, *ring.Ring]
+	ring          *ring.Ring
+	size          int
+}
+
+// NewBufferQueue returns a new BufferQueue.
+func NewBufferQueue(maxBuffer int) *BufferQueue {
+	return &BufferQueue{
+		maxBuffer:     maxBuffer,
+		activeIssuers: shrinkingmap.New[iotago.AccountID, *ring.Ring](),
+		ring:          nil,
+	}
+}
+
+// NumActiveIssuers returns the number of active issuers in b.
+func (b *BufferQueue) NumActiveIssuers() int {
+	return b.activeIssuers.Size()
+}
+
+// MaxSize returns the max number of blocks in BufferQueue.
+func (b *BufferQueue) MaxSize() int {
+	return b.maxBuffer
+}
+
+// Size returns the total number of blocks in BufferQueue.
+func (b *BufferQueue) Size() int {
+	return b.size
+}
+
+// IssuerQueue returns the queue for the corresponding issuer.
+func (b *BufferQueue) IssuerQueue(issuerID iotago.AccountID) *IssuerQueue {
+	element, ok := b.activeIssuers.Get(issuerID)
+	if !ok {
+		return nil
+	}
+	issuerQueue, isIQ := element.Value.(*IssuerQueue)
+	if !isIQ {
+		return nil
+	}
+
+	return issuerQueue
+}
+
+func (b *BufferQueue) GetOrCreateIssuerQueue(issuerID iotago.AccountID) (*IssuerQueue, error) {
+	element, issuerActive := b.activeIssuers.Get(issuerID)
+	if issuerActive {
+		issuerQueue, isIQ := element.Value.(*IssuerQueue)
+		if !isIQ {
+			return nil, ierrors.New("buffer contains elements that are not issuer queues")
+		}
+
+		return issuerQueue, nil
+	}
+	issuerQueue := NewIssuerQueue(issuerID)
+	b.activeIssuers.Set(issuerID, b.ringInsert(issuerQueue))
+
+	return issuerQueue, nil
+}
+
+// Submit submits a block. Return blocks dropped from the scheduler to make room for the submitted block.
+// The submitted block can also be returned as dropped if the issuer does not have enough access mana.
+func (b *BufferQueue) Submit(blk *blocks.Block, manaRetriever func(iotago.AccountID) (iotago.Mana, error)) (elements []*blocks.Block, err error) {
+	issuerID := blk.ProtocolBlock().IssuerID
+	issuerQueue, err := b.GetOrCreateIssuerQueue(issuerID)
+	if err != nil {
+		return nil, ierrors.Errorf("%w: could not get or create issuer queue for issuer %s", err, issuerID)
+	}
+
+	// first we submit the block, and if it turns out that the issuer doesn't have enough bandwidth to submit, it will be removed by dropTail
+	if !issuerQueue.Submit(blk) {
+		return nil, ierrors.Errorf("block already submitted %s", blk)
+	}
+
+	b.size++
+
+	// if max buffer size exceeded, drop from tail of the longest mana-scaled queue
+	if b.Size() > b.maxBuffer {
+		return b.dropTail(manaRetriever), nil
+	}
+
+	return nil, nil
+}
+
+func (b *BufferQueue) dropTail(manaRetriever func(iotago.AccountID) (iotago.Mana, error)) (droppedBlocks []*blocks.Block) {
+	start := b.Current()
+	// remove as many blocks as necessary to stay within max buffer size
+	for b.Size() > b.maxBuffer {
```

Comment:
I think it is ok to drop based on work in the buffers even though we are trying to minimise the size in the buffer. Work and size will be at least correlated strongly, and we will still make the best possible use of the buffer size capacity, although sometime it will not fit as much "work" in it as other times.